src/DataStructures/Index.hpp

```diff
@@ -118,12 +141,75 @@ class Index {
 /// Index. Note that the first dimension of the Index varies fastest when
 /// computing the collapsed index.
 template <size_t N>
-size_t collapsed_index(const Index<N>& index, const Index<N>& extents) noexcept;
+size_t collapsed_index(const Index<N>& index,
+                       const Index<N>& extents) noexcept {
+  return extents.collapsed_index(index.indices_);
```

Comment:
Why are you adding the member function, instead of keeping only the free function? I feel like we should choose one, and I remember you are always in favour of free functions over member functions

Reply:
Mostly to make it easier to find, but that's not a great reason. I'll change it.